daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
 object TransactionCoder {
 
   import ValueCoder.{DecodeCid, EncodeCid, codecContractId}
-  type EncodeNid[Nid] = Nid => String
-  type EncodeVal[Cid <: Value.ContractId] =
-    VersionedValue[Cid] => Either[EncodeError, (ValueVersion, ValueOuterClass.VersionedValue)]
+
+  abstract class EncodeNid[Nid] private[lf] {
+    def asString(id: Nid): String
+  }
+  abstract class DecodeNid[Nid] private[lf] {
+    def fromString(s: String): Either[DecodeError, Nid]
+  }
+
+  val NidEncoder: EncodeNid[Value.NodeId] = new EncodeNid[Value.NodeId] {
```

Comment:
```suggestion
  val NidEncoder: EncodeNid[Value.NodeId] = _.index.toString
```

Reply:
does not work with the `private[lf]`at line 29